test/extensions/filters/http/ext_authz/config_test.cc

```diff
@@ -3,22 +3,177 @@
 #include "envoy/extensions/filters/http/ext_authz/v3/ext_authz.pb.validate.h"
 #include "envoy/stats/scope.h"
 
+#include "source/common/grpc/async_client_manager_impl.h"
+#include "source/common/network/address_impl.h"
+#include "source/common/thread_local/thread_local_impl.h"
 #include "source/extensions/filters/http/ext_authz/config.h"
 
 #include "test/mocks/server/factory_context.h"
 #include "test/test_common/test_runtime.h"
 #include "test/test_common/utility.h"
 
+#include "absl/synchronization/notification.h"
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
 
 using testing::_;
 using testing::Invoke;
+using testing::NiceMock;
 
 namespace Envoy {
 namespace Extensions {
 namespace HttpFilters {
 namespace ExtAuthz {
+
+class MultiThreadTest {
+public:
+  MultiThreadTest(size_t num_threads) : num_threads_(num_threads), api_(Api::createApiForTest()) {}
+  virtual ~MultiThreadTest() { tls_.shutdownGlobalThreading(); }
+  virtual void workerFunc() {}
+  void run() { spawnAllWorkers(); }
+
+private:
+  void spawnAllWorkers() {
+    // Create dispatcher and register.
+    for (size_t i = 0; i < num_threads_; i++) {
+      dispatchers_.emplace_back(api_->allocateDispatcher(std::to_string(i)));
+      tls_.registerThread(*dispatchers_[i], false);
+    }
+    // Create running threads.
+    for (size_t i = 0; i < num_threads_; i++) {
+      // i must be explicitly captured by value.
+      workers_.emplace_back(api_->threadFactory().createThread(
+          [&, i]() { dispatchers_[i]->run(Event::Dispatcher::RunType::RunUntilExit); }));
+    }
+    // Post callback to fire all threads at the same time.
+    for (Event::DispatcherPtr& dispatcher : dispatchers_) {
+      dispatcher->post([&]() {
+        waitForAllWorkersToFire();
+        workerFunc();
+      });
+    }
+    // Post exit signals and wait for thread to end.
+    for (Event::DispatcherPtr& dispatcher : dispatchers_) {
+      dispatcher->post([&dispatcher]() { dispatcher->exit(); });
+    }
+    for (Thread::ThreadPtr& worker : workers_) {
+      worker->join();
+    }
+  }
+
+  void waitForAllWorkersToFire() {
+    if (shouldFire()) {
+      // The last thread spawned should notify all to start.
+      workers_should_fire_.Notify();
+    } else {
+      // Wait for notification if the current thread is not the last one spawned.
+      workers_should_fire_.WaitForNotification();
+    }
+  }
+
+  bool shouldFire() {
+    Thread::LockGuard lock(mutex_);
+    active_threads_cnt_++;
+    return active_threads_cnt_ == num_threads_;
+  }
+
+  size_t num_threads_;
+  absl::Notification workers_should_fire_;
+  Thread::MutexBasicLockable mutex_;
+  size_t active_threads_cnt_{};
```

Comment:
ABSL_GUARDED_BY(mutex_)

...though that won't be needed if you use a blocking counter.